File: sortingAlgorithms.js
function quickSort(arr, low , high){
    if(low < high){
        let pi = partition(arr, low, high);
        quickSort(arr, low, pi-1);
        quickSort(arr, pi+1, high);
    }
}

function bubbleSort(arr){
    let n = arr.length;
    for(let i = 0; i < n-1; i++){
        for(let j = 0; j < n-i-1; j++){
            if(arr[j]> arr[j+1]){
                let temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }}
}

function mergeSort(arr, low , high){
    if(low < high){
        let mid =Math.floor ((high+low)/2);
        mergeSort(arr,low,mid )
        mergeSort(arr,mid+1, high)
        merge(arr , low , mid , high)
}
}   

function insertSort(arr){
    for (let i = 1; i < arr.length; i++) {
        let key = arr[i];
        let j = i-1;
        while (j >= 0 && key < arr[j]) {
            arr[j+1]= arr[j];
            j--;
        }
    arr[j+1]=key;
    }
}

function swap (arr , i , j ){
    let temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function merge(arr, low, mid, high) {
    let arr1 = new Array(mid - low + 1); // Temporary array 1
    let arr2 = new Array(high - mid);    // Temporary array 2

    // Copy elements into temporary arrays
    for (let i = 0; i < arr1.length; i++) {
        arr1[i] = arr[low + i];
    }
    for (let i = 0; i < arr2.length; i++) {
        arr2[i] = arr[mid + 1 + i];
    }

    let i = 0, j = 0, k = low;

    // Merge the two arrays back into the original array
    while (i < arr1.length && j < arr2.length) {
        if (arr1[i] <= arr2[j]) {
            arr[k] = arr1[i];
            i++;
        } else {
            arr[k] = arr2[j];
            j++;
        }
        k++;
    }

    // Copy any remaining elements from arr1
    while (i < arr1.length) {
        arr[k] = arr1[i];
        i++;
        k++;
    }

    // Copy any remaining elements from arr2
    while (j < arr2.length) {
        arr[k] = arr2[j];
        j++;
        k++;
    }
}
function partition(arr, low, high) {
    let pivot = arr[low]; // Pivot is the first element
    let i = low + 1;
    let j = high;

    while (i <= j) {
        // Move `i` to the right until we find an element greater than the pivot
        while (i <= j && arr[i] <= pivot) {
            i++;
        }

        // Move `j` to the left until we find an element smaller than or equal to the pivot
        while (i <= j && arr[j] > pivot) {
            j--;
        }

        // Swap out-of-place elements
        if (i < j) {
            let temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
        }
    }

    // Place the pivot in its correct position
    let temp = arr[low];
    arr[low] = arr[j];
    arr[j] = temp;

    return j; // Return the index of the pivot
}



let arr = [5,4,9,32,8,5];
let n = arr.length;
let low = 0;
let high = n-1;
quickSort(arr, low , n-1)
console.log(arr);

File: sortingAlgorithms.py
def quick_sort(arr, low, high):
    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)

def bubble_sort(arr):
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                temp = arr[j]
                arr[j] = arr[j + 1]
                arr[j + 1] = temp

def merge_sort(arr, low, high):
    if low < high:
        mid = (high + low) // 2
        merge_sort(arr, low, mid)
        merge_sort(arr, mid + 1, high)
        merge(arr, low, mid, high)

def insert_sort(arr):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

def swap(arr, i, j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp

def merge(arr, low, mid, high):
    # Create temporary arrays
    arr1 = [0] * (mid - low + 1)
    arr2 = [0] * (high - mid)

    # Copy data to temporary arrays
    for i in range(mid - low + 1):
        arr1[i] = arr[low + i]
    for i in range(high - mid):
        arr2[i] = arr[mid + 1 + i]

    # Merge the temporary arrays back into the original array
    i = 0
    j = 0
    k = low

    while i < len(arr1) and j < len(arr2):
        if arr1[i] <= arr2[j]:
            arr[k] = arr1[i]
            i += 1
        else:
            arr[k] = arr2[j]
            j += 1
        k += 1

    # Copy any remaining elements from arr1
    while i < len(arr1):
        arr[k] = arr1[i]
        i += 1
        k += 1

    # Copy any remaining elements from arr2
    while j < len(arr2):
        arr[k] = arr2[j]
        j += 1
        k += 1

def partition(arr, low, high):
    pivot = arr[low]
    i = low + 1
    j = high

    while i <= j:
        # Find an element on the left that should be on the right
        while i <= j and arr[i] <= pivot:
            i += 1

        # Find an element on the right that should be on the left
        while i <= j and arr[j] > pivot:
            j -= 1

        # Swap out-of-place elements
        if i < j:
            arr[i], arr[j] = arr[j], arr[i]

    # Place the pivot in its correct position
    arr[low], arr[j] = arr[j], arr[low]
    return j  # Return the final pivot position


arr = [5, 4, 9, 32, 8, 5]
n = len(arr)
low = 0 
high = n - 1
merge_sort(arr,low,high)
print(arr)
